### PR TITLE
Plugin manager clean up

### PR DESCRIPTION
**Pull request by Bombe**
Lots of clean ups and refactorings in one of `PluginManager`’s most ugly methods.

### CODE REVIEW

File: src/freenet/pluginmanager/PluginManager.java [hunk 1]
```diff
@@ -324,12 +308,14 @@ private static String pluginList(ArrayList<PluginInfoWrapper> wrappers) {
 	}
 
 	private String[] getConfigLoadString() {
-		ArrayList<String> v = new ArrayList<String>();
+		List<String> v = new ArrayList<String>();
+		for (PluginInfoWrapper pi : pluginWrappers) {
```
Comment (unixninja92) [inline src/freenet/pluginmanager/PluginManager.java]: Why did you move this before the if(!started) check? I understand why you moved it out of the synchronized block, but if it's not started, why do you want to bother filling in v? Might that have some effect on the timing of wether started is true or not?

File: src/freenet/pluginmanager/PluginManager.java [hunk 2]
```diff
@@ -1064,65 +975,35 @@ public String handleHTTPPost(String plugin, HTTPRequest request) throws PluginHT
...
-		boolean found = false;
-		synchronized(pluginWrappers) {
-			for(int i = 0; i < pluginWrappers.size() && !found; i++) {
-				pi = pluginWrappers.get(i);
-				if(pi.getThreadName().equals(name)) {
-					found = true;
-					break;
-				}
+		for (PluginInfoWrapper pluginInfoWrapper : pluginWrappers) {
+			if (pluginInfoWrapper.getThreadName().equals(name)) {
```
Comment (unixninja92) [inline src/freenet/pluginmanager/PluginManager.java]: Why did you remove the break after stopping the plugin? I don't see why we should keep looking through threads.

Comment (unixninja92) [inline src/freenet/pluginmanager/PluginManager.java]: Same question for the other killPlugin methods

File: src/freenet/pluginmanager/PluginManager.java [hunk 3]
```diff
@@ -1183,266 +1064,246 @@ public File getPluginFilename(String pluginName) {
...
 	 * @throws PluginNotFoundException
 	 *             If anything goes wrong.
+	 * @throws PluginAlreadyLoaded if the plugin is already loaded
 	 */
-	private FredPlugin loadPlugin(PluginDownLoader<?> pdl, String name, PluginProgress progress, boolean ignoreOld) throws PluginNotFoundException {
+	private FredPlugin loadPlugin(PluginDownLoader<?> pdl, String name, PluginProgress progress, boolean alwaysDownload) throws PluginNotFoundException, PluginAlreadyLoaded {
 
 		pdl.setSource(name);
 
-		/* check for plugin directory. */
```
Comment (unixninja92) [inline src/freenet/pluginmanager/PluginManager.java]: Why are we removing this check?

**Comment by unixninja92** [comment]
@Thynix reviewed up to ```private FredPlugin loadPlugin```

File: src/freenet/pluginmanager/PluginManager.java (hunk 1 above)
Comment (Bombe) [inline src/freenet/pluginmanager/PluginManager.java]: Mostly because if I didn’t I would either have to let it stay in the `synchronized` block or would have to split up the `synchronized` block into two `synchronized` blocks… I’m not sure which solution is better. :)

File: src/freenet/pluginmanager/PluginManager.java (hunk 2 above)
Comment (Bombe) [inline src/freenet/pluginmanager/PluginManager.java]: Hmm, no reason, really. Re-added.

File: src/freenet/pluginmanager/PluginManager.java [hunk 4]
```diff
@@ -280,41 +269,36 @@ public void stop(long maxWaitTime) {
 			progress.kill();
 		}
 		// Stop already loaded plugins.
-		ArrayList<PluginInfoWrapper> wrappers;
-		synchronized(pluginWrappers) {
-			wrappers = new ArrayList<PluginInfoWrapper>(pluginWrappers);
-		}
-		for(PluginInfoWrapper pi : wrappers)
+		for(PluginInfoWrapper pi : pluginWrappers)
```
Comment (bertm) [inline src/freenet/pluginmanager/PluginManager.java]: This should have braces.

File: src/freenet/pluginmanager/PluginManager.java (hunk 1 above)
Comment (bertm) [inline src/freenet/pluginmanager/PluginManager.java]: Doing it this way has the potential (however unlikely) to return a wrong result due to a race condition where the snapshots of `pluginWrappers` and `pluginsFailedLoad`, not taken atomically as a pair, are inconsistent.

Including a little extra work in the `synchronized` block is not that bad, we don't expect this method to be called too often anyhow.

File: src/freenet/pluginmanager/PluginManager.java [hunk 5]
```diff
@@ -409,17 +395,17 @@ private PluginInfoWrapper realStartPlugin(final PluginDownLoader<?> pdl, final S
...
-			if (plug == null)
-				return null; // Already loaded
-			pluginProgress.setProgress(PluginProgress.PROGRESS_STATE.STARTING);
+			plug = loadPlugin(pdl, filename, pluginProgress, alwaysDownload);
+			pluginProgress.setProgress(ProgressState.STARTING);
 			pi = new PluginInfoWrapper(node, plug, filename, pdl.isOfficialPluginLoader());
 			PluginHandler.startPlugin(PluginManager.this, pi);
+			pluginWrappers.add(pi);
 			synchronized (pluginWrappers) {
-				pluginWrappers.add(pi);
```
Comment (bertm) [inline src/freenet/pluginmanager/PluginManager.java]: This should remain in the `synchronized` block, for `pluginWrappers` and `pluginsFailedLoad` to remain consistent.

File: src/freenet/pluginmanager/PluginManager.java [hunk 6]
```diff
@@ -1183,266 +1068,246 @@ public File getPluginFilename(String pluginName) {
...
 		/* check if file needs to be downloaded. */
 		if(logMINOR)
 			Logger.minor(this, "plugin file " + pluginFile.getAbsolutePath() + " exists: " + pluginFile.exists()+" downloader "+pdl+" name "+name);
 		int RETRIES = 5;
-		for(int i = 0; i < RETRIES; i++) {
-			if(!pluginFile.exists() || pluginFile.length() == 0)
+		for (int i = 0; i < RETRIES; i++) {
+			if (!pluginFile.exists() || pluginFile.length() == 0) {
 				try {
 					downloaded = true;
```
Comment (bertm) [inline src/freenet/pluginmanager/PluginManager.java]: Why is this flag set here? It's clearly not downloaded at this point and the flag will never be reset to `false`.

File: src/freenet/pluginmanager/PluginManager.java [hunk 7]
```diff
@@ -1183,266 +1068,246 @@ public File getPluginFilename(String pluginName) {
...
+	private void verifyDigest(PluginDownLoader<?> pluginDownLoader, File pluginFile) throws PluginNotFoundException {
+		String digest = pluginDownLoader.getSHA1sum();
+		if (digest == null) {
+			return;
+		}
+		String testsum = getFileDigest(pluginFile, "SHA-1");
+		if (!(digest.equalsIgnoreCase(testsum))) {
+            Logger.error(this, "Checksum verification failed, should be " + digest + " but was " + testsum);
+            throw new PluginNotFoundException("Checksum verification failed, should be " + digest + " but was " + testsum);
+        }
```
Comment (bertm) [inline src/freenet/pluginmanager/PluginManager.java]: This whitespace is inconsistent.

File: src/freenet/pluginmanager/PluginManager.java [hunk 8]
```diff
@@ -820,66 +806,16 @@ public void unregisterPluginToadlet(PluginInfoWrapper pi) {
 		}
 	}
 
-	public void addToadletSymlinks(PluginInfoWrapper pi) {
```
Comment (bertm) [inline src/freenet/pluginmanager/PluginManager.java]: Please deprecate this first, so we won't break plugins unnecessarily.

File: src/freenet/pluginmanager/PluginManager.java [hunk 9]
```diff
@@ -820,66 +806,16 @@ public void unregisterPluginToadlet(PluginInfoWrapper pi) {
...
-					Logger.normal(this, "Removed HTTP symlink: " + target +
-						" => /plugins/" + pi.getPluginClassName() + '/');
-				}
-			} catch(Throwable ex) {
-				Logger.error(this, "removing Toadlet-link", ex);
-			}
-		}
-	}
-
-	public void removeToadletSymlinks(PluginInfoWrapper pi) {
```
Comment (bertm) [inline src/freenet/pluginmanager/PluginManager.java]: As above, please deprecate this first.

File: src/freenet/pluginmanager/PluginManager.java [hunk 10]
```diff
@@ -1706,39 +1559,35 @@ public static void setLanguage(LANGUAGE lang) {
...
+						} catch (Throwable t) {
+							Logger.error(this, "Cought Trowable in Callback", t);
+						}
+					}
+				}, "Callback");
 			}
 		}
 	}
 
-	public THEME getFProxyTheme() {
```
Comment (bertm) [inline src/freenet/pluginmanager/PluginManager.java]: Please deprecate this before removing it entirely.

File: src/freenet/pluginmanager/PluginManager.java (hunk 4 above)
Comment (Bombe) [inline src/freenet/pluginmanager/PluginManager.java]: It should. It will.

File: src/freenet/pluginmanager/PluginManager.java (hunk 1 above)
Comment (Bombe) [inline src/freenet/pluginmanager/PluginManager.java]: If they need to be synchronized together, maybe they should be their own class.

File: src/freenet/pluginmanager/PluginManager.java (hunk 6 above)
Comment (Bombe) [inline src/freenet/pluginmanager/PluginManager.java]: The point of this flag is not to shows that the plugin has already been successfully downloaded, it’s for showing that an attempt to download was made. On the first run of the loop the plugin may already exist; errors on loading will then trigger a re-download. Only if this run (with `i = 1`) fails, an exception is thrown.

So, yes, the flag’s name is stupid, and it took me a while to understand. And yes, its name should be better. Will fix.

File: src/freenet/pluginmanager/PluginManager.java (hunk 7 above)
Comment (Bombe) [inline src/freenet/pluginmanager/PluginManager.java]: It indeed is. Will fix.

File: src/freenet/pluginmanager/PluginManager.java (hunk 10 above)
Comment (Bombe) [inline src/freenet/pluginmanager/PluginManager.java]: Okay, though I think that plugins that ask the plugin manager for the current fproxy theme should be taken out back and shot. Should we schedule a date for removal? In six months, maybe? A year?

File: src/freenet/pluginmanager/PluginManager.java (hunk 8 above)
Comment (Bombe) [inline src/freenet/pluginmanager/PluginManager.java]: I did a search with grep in my work directory and even though I don’t claim to have all plugins’ source code I got no hits in anywhere but the plugin manager. I will deprecate them but I would like to schedule date for removal.

File: src/freenet/pluginmanager/PluginManager.java (hunk 9 above)
Comment (Bombe) [inline src/freenet/pluginmanager/PluginManager.java]: ACK.

File: src/freenet/pluginmanager/PluginManager.java (hunk 8 above)
Comment (Bombe) [inline src/freenet/pluginmanager/PluginManager.java]: Actually, take a look at this method: it does not even pretend to do what it’s named for. I’m more in favor of simply removing this method before anybody notices its complete brokenness. :)

**Comment by bertm** [comment]
Given the state of the methods now deprecated, I suggest we remove them entirely in 1473.

**Comment by bertm** [comment]
As I see it, `startingPlugins`, `pluginWrappers` and `pluginsFailedLoad` should be kept consistent. Mostly for presentation purposes: if we fail doing so, the downside would be the possibility of a condition where the user would temporarily be presented with incorrect information on the plugin state. There are a couple more issues with synchronization (such as access to `startingPlugins` being synchronized on either `startingPlugins`, `this` or not at all) that will become apparent when doing so.

This pull request does not appear to make things worse than they were before. So if you'd rather not do that at this time, it can be done in a separate pull request.

File: src/freenet/pluginmanager/PluginManager.java (hunk 8 above)
Comment (bertm) [inline src/freenet/pluginmanager/PluginManager.java]: Yes, it's a great example of copy-pasting gone wrong ;)

**Comment by Bombe** [comment]
@bertm I agree on keeping those three together. I am already working on something for that which should be finished in a couple of hours, hopefully. Also, I am completely fine with removing all those now-deprecated methods in 1473.

**Comment by bertm** [comment]
Apparently I missed the additional progress made.

This looks good to me so far, but  we have merge conflicts now. @Bombe will you resolve those, or should I do it?

**Comment by xor-freenet** [comment]
@Bombe It seems bertm has ACKed this a long time ago and it merely lacks merging of next into it. Can you do that?

**Comment by ArneBab** [comment]
@Bombe this has been marked as "after next release" for quite some time. I’d say with 1483 having only few regressions compared to 1480, this qualifies as after next release. Once you get to fix the conflicts, I’d like to merge this.

**Comment by Bombe** [comment]
> @Bombe this has been marked as "after next release" for quite some time. I’d say with 1483 having only few regressions compared to 1480, this qualifies as after next release. Once you get to fix the conflicts, I’d like to merge this.

I had some more commits lying around in that branch with some more minor fixes; they all have been rebased on `next` so should be good to go.

**Comment by ArneBab** [comment]
Cool! Thank you!
